braindecode/augmentation/functional.py

```diff
@@ -969,3 +970,115 @@ def mixup(X, y, lam, idx_perm):
         y_b[idx] = y[idx_perm[idx]]
 
     return X_mix, (y_a, y_b, lam)
+
+
+def _segmentation_reconstruction(X, n_segments, rng):
+    n_trials, n_channels, window_size = X.shape
+    # Segment Size
+    segment_size = window_size // n_segments
+    # Initialize an empty tensor for augmented data
+    X_aug = torch.zeros_like(X)
+    # Generate random indices within the class-specific dataset
+    rand_idx = rng.randint(0, n_trials, (n_trials, n_segments))
+    for idx_segment in range(n_segments):
+        start = idx_segment * segment_size
+        end = (idx_segment + 1) * segment_size
+
+        # Perform the data augmentation
+        X_aug[np.arange(n_trials), :, start:end] = X[
+            rand_idx[:, idx_segment], :, start:end
+        ]
+    return X_aug, n_trials
+
+
+def segmentation_reconstruction(X, y, n_segments=None, random_state=None):
+    """Segment and reconstruct EEG data.
+
+    See [1]_ for details.
+
+    Parameters
+    ----------
+    X : torch.Tensor
+        EEG input example or batch.
+    y : torch.Tensor
+        EEG labels for the example or batch.
+    n_segments : int
+        Number of segments to use in the batch. If None, X will be
+        automatically segmented, getting the last element in a list
+        of factors of the number of samples's square root. Defaults to None.
+    random_state: int | numpy.random.Generator, optional
+        Used to draw the phase perturbation. Defaults to None.
+    Returns
+    -------
+    torch.Tensor
+        Transformed inputs.
+    torch.Tensor
+        Transformed labels.
+    References
+    ----------
+    .. [1] Lotte, F. (2015). Signal processing approaches to minimize or
+    suppress calibration time in oscillatory activity-based brain–computer
+    interfaces. Proceedings of the IEEE, 103(6), 871-890.
+    """
+
+    if y is not None:
+        if not isinstance(X, torch.Tensor) or not isinstance(y, torch.Tensor):
+            raise ValueError("X and y must be torch tensors.")
+
+        if X.shape[0] != y.shape[0]:
+            raise ValueError("Number of samples in X and y must be the same.")
+
+    if n_segments is None:
+        n_segments = int(X.shape[2])
+        n_segments_list = []
+        for i in range(1, int(n_segments**0.5) + 1):
+            if n_segments % i == 0:
+                n_segments_list.append(i)
+        n_segments = n_segments_list[-1]
```

Comment:
I find the construction of n_segments curious here when the kwarg is None. Could you explain or provide a reference which explains this please, just for my own culture ?

Reply:
Hi @cedricrommel,

The choice for the default value came from some tests I did in a recent work (yet to be published) in which I use this type of augmentation and also some results I obtained while testing the integration of the technique in braindecode. Firstly, the factors of the number of samples seem to perform better than random values (since the segmentation forms a trial with the right size, without the need for padding). Secondly, in the datasets I tested (mainly BNCI2014_001), a “reasonable” number of segments (close to the square root of the number of samples) seems to have a better result than too many segments (probably due to the loss of coherence with so many different segments coming together to form a single trial). I therefore chose to use the maximum number of segments within this “reasonable” range for a general context. We can do more tests in the future to find a better default value, if possible.